### DISCUSSION TITLE
Avoid orphaned entries in M2M relations (set null) and delete instead

### DISCUSSION THREAD

**Discussion by nickluger** [directus/directus#5149] [Feature Requests]
If we have an M2M relation, and we disassociate one of the entries on one side of the relation, Directus sets the "foreign key" on the relation to `null` via an update. It does not delete the row. But it doesn't re-use it neither, if we re-associate the same entry again. This leads to an **orphaned row**, having a `null` value on one of the foreign keys but having no obvious purpose any longer.

We should think about deleting M2M-table-rows instead of setting foreign keys to `null`. Deleting the row - which actually represents this M2M-relationship - would be the conceptually right thing to do IMHO. Is there a reason why we use the update here? Are these orphaned rows needed for something?

I know that Directus does not use the relational DB concept of foreign keys currently (which is a pity), but I'm using the term here as it's used in the docs.

If there's an issue I overlooked or an already open discussion, please point me to that. Thanks! 🙂

**Comment by rijkvanzanten** [comment]
> If there's an issue I overlooked or an already open discussion, please point me to that. Thanks! 🙂

https://github.com/directus/directus/issues/2942

There you go! 😄

**Reply by nickluger** [reply]
Thanks @rijkvanzanten , i knew the pinned cascade issue 😄

I'm just not sure why this is an fkey/cascade issue, as the row **containing the fkeys** (the row in the middle of the M2M) should be deleted when disassociating. Cascade delete would mean that a row is deleted as the result of the deletion of a row referenced by an fkey. But there's no deletion in this case, so no cascade.  

https://github.com/directus/directus/issues/2942 would only be applicable, if i deleted an entry on **one side** of the relation, then this deletion should trigger a cascade delete on the entry in the M2M-table, too. 

But here we need to call `delete` explicitly (instead of `update`) on the M2M relation itself.

**Reply by rijkvanzanten** [reply]
Yeah, #2942 will incorporate both 👍🏻 The foreign key constraints are DB native, but we'll also have our own "nullify vs delete" for o2m setups 👍🏻

**Reply by nickluger** [reply]
Perfect 👌
